docs/en/integrations/language-clients/java/index.md

```diff
@@ -5,126 +5,250 @@ keywords: [clickhouse, java, jdbc, client, integrate, r2dbc]
-- `http://(https://explorer@play.clickhouse.com:443`
-- `tcp://localhost?!auto_discovery#experimental),(grpc://localhost#experimental)?failover=3#test`
+All settings are defined by instance methods (a.k.a configuration methods) that make scope and context of each value clear. 
+Major configuration parameters are defined in one scope (client or operation) and do not override each other. Handling 
+configuration overriding across is a very hard task so we doing our best to keep it simple. 
 
+This section describes only client wide settings. Each operation may have own and will be listed in the their sections. 
 
-Connect to a cluster with multiple nodes:
 
-```java
-ClickHouseNodes servers = ClickHouseNodes.of(
-    "jdbc:ch:http://server1.domain,server2.domain,server3.domain/my_db"
-    + "?load_balancing_policy=random&health_check_interval=5000&failover=2");
+### Insert API
+
+| Method                                                             | Description |
+|--------------------------------------------------------------------|-------------|
+| `CompletableFuture<InsertResponse> insert(String tableName, InputStream data, ClickHouseFormat format, InsertSettings settings)` | Sends write request to database. Input data is read from the input stream. |
+| `CompletableFuture<InsertResponse> insert(String tableName, List<?> data, InsertSettings settings)` | Sends write request to database. List of objects is converted into a most effective format and then is sent to a server. Class of the list items should be registed up-front using `register(Class, TableSchema)` method.
+
+There is no much difference in performance between raw data insert and POJO list insert. The last one helps to avoid 
+coding complex serialization into RowBinary format - client will handle it for you.   
+
+```java 
+// Important step (done once) - register class to pre-compile object serializer according to the table schema. 
+client.register(ArticleViewEvent.class, client.getTableSchema(TABLE_NAME));
+
+
+List<ArtivleViewEvent> events = loadBatch();
+
+try (InsertResponse response = client.insert(TABLE_NAME, events).get()) {
+    // handle response, then it will be closed and connection that served request will be released. 
+}
 ```
 
-Connect to a single node:
+If data is already in the format that ClickHouse accepts, then: 
+```java 
+try (InputStream dataStream = getDataStream()) {
+    try (InsertResponse response = client.insert(TABLE_NAME, dataStream, ClickHouseFormat.JSONEachRow,
+            insertSettings).get(3, TimeUnit.SECONDS)) {
+
+        log.info("Insert finished: {} rows written", response.getMetrics().getMetric(ServerMetrics.NUM_ROWS_WRITTEN).getLong());
+    } catch (Exception e) {
+        log.error("Failed to write JSONEachRow data", e);
+        throw new RuntimeException(e);
+    }
+}
 
-```java
-ClickHouseNode server = ClickHouseNode.of("http://localhost:8123/default?compress=0");
 ```
 
-### Compression
+### Query API
```

Comment:
I would expand the section with more specific examples: create a table, select, drop, etc. It can be done in a follow-up

Reply:
Expanded. Will add more later